provider.go

```diff
@@ -718,8 +728,78 @@ func (p *provider) StreamInvoke(*rpc.InvokeRequest, rpc.ResourceProvider_StreamI
 	return status.Error(codes.Unimplemented, "StreamInvoke is not yet implemented")
 }
 
-func (p *provider) Call(context.Context, *rpc.CallRequest) (*rpc.CallResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "Call is not yet implemented")
+func (p *provider) Call(ctx context.Context, req *rpc.CallRequest) (*rpc.CallResponse, error) {
+
+	configPropertyMap := make(presource.PropertyMap, len(req.GetConfig()))
+	for k, v := range req.GetConfig() {
+		configPropertyMap[presource.PropertyKey(k)] = presource.NewProperty(v)
+	}
+	pulumiContext, err := pulumi.NewContext(ctx, pulumi.RunInfo{
+		Project:           req.GetProject(),
+		Stack:             req.GetStack(),
+		Config:            req.GetConfig(),
+		ConfigSecretKeys:  req.GetConfigSecretKeys(),
+		ConfigPropertyMap: configPropertyMap,
+		Parallel:          int(req.GetParallel()),
+		DryRun:            req.GetDryRun(),
+		MonitorAddr:       req.GetMonitorEndpoint(),
+		Organization:      req.GetOrganization(),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to build pulumi.Context: %w", err)
+	}
+
+	args, err := p.getMap(req.GetArgs())
+	if err != nil {
+		return nil, fmt.Errorf("unable to convert args into a property map")
+	}
+
+	resp, err := p.client.Call(ctx, CallRequest{
+		Tok:     tokens.ModuleMember(req.GetTok()),
+		Args:    args,
+		Context: pulumiContext,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	returnDependencies := map[string]*rpc.CallResponse_ReturnDependencies{}
+	for name, v := range resp.Return {
+		var urns []string
+		resourcex.Walk(v, func(v presource.PropertyValue, state resourcex.WalkState) {
+			if state.Entering || !v.IsOutput() {
+				return
+			}
+			for _, dep := range v.OutputValue().Dependencies {
+				urns = append(urns, string(dep))
+			}
+		})
+
+		returnDependencies[string(name)] = &rpc.CallResponse_ReturnDependencies{Urns: urns}
```

Comment:
Is this needed even when the provider has first-class outputs in Return, where this approximation can be un-abmiguously computed? Probably no harm in including but I thought maybe engine does this?

Reply:
I'm not sure, but there is no documentation saying that it isn't necessary.

It seems like it is not necessary in Pulumi:

https://github.com/pulumi/pulumi/blob/cbfea508e7b2a4f1dd0985cefbeda3a6e891f7f9/pkg/resource/deploy/source_eval.go#L1140-L1142

I'm hesitant to assume that it is never necessary.

Reply:
I opened https://github.com/pulumi/pulumi/pull/17113 to pin down an answer here.